lib/twitter/accounts.ex

```diff
@@ -0,0 +1,74 @@
+defmodule Twitter.Accounts do
+  @moduledoc """
+  The Accounts context.
+  """
+  alias Twitter.Repo
+  alias Twitter.Accounts.User
+
+  @doc """
+  create user: validate parameters and insert in the db
+  """
+  @spec create_user(map()) ::
+          {:ok, User.t()} | {:error, Ecto.Changeset.t(User.t())}
+  def create_user(attrs \\ %{}) do
+    %User{}
+    |> User.registration_changeset(attrs)
+    |> Repo.insert()
+  end
+
+  @doc """
+  get_user: returns user by id
+  """
+  @spec get_user(non_neg_integer() | String.t()) :: User.t() | nil
+  def get_user(id) do
+    Repo.get(User, id)
+  end
+
+  @doc """
+  returns user by id or raises Ecto.NoResultsError if not found
+  """
+  @spec get_user!(non_neg_integer() | String.t()) :: User.t()
+  # has error handling
+  def get_user!(id) do
+    Repo.get!(User, id)
+  end
+
+  @doc """
+  returns a single user by any parameter
+  """
+  @spec get_user_by(map()) :: User.t() | nil
+  def get_user_by(params) do
+    Repo.get_by(User, params)
+  end
+
+  @doc """
+  lists all users in db
+  """
+  @spec list_users :: list(User.t())
+  def list_users do
+    Repo.all(User)
+  end
+
+  @doc """
+  find user by username, compare input password to the stored one.
+  returns {:ok, user} on success, {:error, :errorcode} on failure
+  """
+  @spec authenticate_by_username_and_pass(String.t(), String.t()) ::
+          {:ok, User.t()} | {:error, :unauthorized | :not_found}
+
+  def authenticate_by_username_and_pass(username, given_pass) do
+    user = get_user_by(username: username)
```

Comment:
you can call `Repo.get_by(User, username: username)` directly here

Reply:
ok!